drivers: fuelguage: Add ADI LTC2959 driver

LostinTimeandspaceYT opened the pull request:
> The LTC®2959 is an ultra-low power battery gas gauge
that accurately measures charge, voltage, current and
temperature. Its wide input voltage range and its low
operating current make the LTC2959 suitable for many
applications, including duty-cycled systems that operate
over long lifetime.

This driver integrates with Zephyr’s fuel gauge subsystem to provide access to telemetry data such as voltage, current, temperature, and accumulated charge.

Tested on custom hardware with known RSENSE configuration and verified against datasheet behavior.

Signed-off-by: Nathan Winslow <natelostintimeandspace@gmail.com>

fmoessbauer commented on drivers/fuel_gauge/ltc2959/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+zephyr_library()
+zephyr_include_directories(${CMAKE_CURRENT_SOURCE_DIR})
+zephyr_library_sources(ltc2959.c)
```
It looks like the driver itself is missing.

fmoessbauer commented:
IMHO this should go after #90310 (or at least 12a8422840071ae3e89a261be5e2d85cac74e535) so things can be tested with the generic example.

ttmut commented on drivers/fuel_gauge/ltc2959/Kconfig:
```diff
@@ -0,0 +1,25 @@
+
+menu "LTC2959 Configuration"
+
+config FUEL_GAUGE_LTC2959
+
+	depends on DT_HAS_ADI_LTC2959_ENABLED
+	bool "LTC2959 Fuel Gauge"
+	default y
+	select I2C
+	help
+	  Enable the LTC2959 fuel gauge driver from Analog Devices.
```
```suggestion
config FUEL_GAUGE_LTC2959
	depends on DT_HAS_ADI_LTC2959_ENABLED
	bool "LTC2959 Fuel Gauge"
	default y
	select I2C
	help
	  Enable the LTC2959 fuel gauge driver from Analog Devices.
```

ttmut commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,575 @@
+	const struct ltc2959_config *cfg = dev->config;
+	mask &= LTC2959_CC_WRITABLE_MASK;
+	mask |= LTC2959_CC_RESERVED_VALUE;
+	LOG_INF("setting cc to: 0x%02X", mask);
```
```suggestion
	LOG_DBG("setting cc to: 0x%02X", mask);
```

LostinTimeandspaceYT commented on drivers/fuel_gauge/ltc2959/Kconfig:
```diff
@@ -0,0 +1,25 @@
+	  Enable the LTC2959 fuel gauge driver from Analog Devices.
+
+
+config FUEL_GAUGE_LTC2959_RSENSE_MOHMS
+
+	depends on FUEL_GAUGE_LTC2959
+	int "Sense resistor value in milliohms"
+	default 80
+	help
+	  Set the value of the RSENSE resistor in milliohms.
+	  This is used to calculate the current and charge resolution.
```
Agreed, I will move this into the binding.

LostinTimeandspaceYT commented on dts/bindings/fuel-gauge/adi,ltc2959.yaml:
```diff
@@ -0,0 +1,24 @@
+  reg:
+    required: true
+
+  label:
+    required: true
```
It was being used in test code to verify the binding. I will remove this property

LostinTimeandspaceYT commented:
Thank you @ttmut  for reviewing this PR. I will have time later this week to work on the needed changes and consolidate the commits as per the guidelines.

DBS06 commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,575 @@
+		raw = (value_uv << 16) / LTC2959_GPIO_UNIPOLAR_UV_SF;
+		break;
+
+	case LTC2959_GPIO_MODE_ALERT:
+	case LTC2959_GPIO_MODE_CHGCOMP:
```
```suggestion
```
If they are doing nothing, the `default:` branch is enough.

DBS06 commented on drivers/fuel_gauge/ltc2959/ltc2959.h:
```diff
@@ -0,0 +1,187 @@
+
+enum ltc2959_custom_props {
+	FUEL_GAUGE_ADC_MODE = FUEL_GAUGE_CUSTOM_BEGIN, //0x8000
+	FUEL_GAUGE_VOLTAGE_HIGH_THRESHOLD,
+	FUEL_GAUGE_VOLTAGE_LOW_THRESHOLD,
+	FUEL_GAUGE_CURRENT_HIGH_THRESHOLD,
+	FUEL_GAUGE_CURRENT_LOW_THRESHOLD,
+	FUEL_GAUGE_TEMPERATURE_HIGH_THRESHOLD,
+	FUEL_GAUGE_TEMPERATURE_LOW_THRESHOLD,
```
For `FUEL_GAUGE_VOLTAGE_LOW_THRESHOLD` the `FUEL_GAUGE_LOW_VOLTAGE_ALARM` already exists.  
But nevertheless, I think it is valid to add the other 5 properties to the `fuel_gauge.h` interface, for me it seems legit to add them, because if we already have a low-voltage-alarm why not have a high-voltage-alarm.

LostinTimeandspaceYT replied:
I elected not to use `FUEL_GAUGE_LOW_VOLTAGE_ALARM` as that reports a percentage of charge. The purpose of these properties is for configuring bounds to trigger a fault in the status register. I am not sure how other ICs of this ilk operate, but I agree @DBS06, I think adding them would make sense.

LostinTimeandspaceYT replied:
@DBS06 You're right I did, classic off-by-1 :sweat_smile: 
<img width="1042" height="91" alt="image" src="https://github.com/user-attachments/assets/db1f296f-ca06-4684-ba15-0857fbca58e2" />

I could change the current `FUEL_GAUGE_LOW_XXXX_THRESHOLD` to `FUEL_GAUGE_LOW_XXX_ALARM` to have the property names be consistent with the current API.

LostinTimeandspaceYT replied:
I started work on this driver in a project on `v4.1.0` and it looks like the `FUEL_GAUGE_LOW_VOLTAGE_ALARM` property was added after that release. I noticed the discrepancy when attempting to build in that project.  I Apologize for the confusion.

DBS06 replied:
You don't need to apologize! That's why code reviews are for.

Yes, I added it after my last PR #90310, which went bigger than originally expected 🤔

Just a minor hint, if you are adding additional properties do that in a separate commit (look at the commits from here #90310) and it should be the first one, as far as I see you can squash all other commits into one commit. So basically in the end you have two commits which will be pushed on the main branch.

I hope this helps!

DBS06 reviewed (changes requested):
First of all, thanks for your contribution and overall it looks well done!   
I also looked inside the datasheet and as far as I see you implemented every necessary/interesting feature from the IC. Interestingly, the IC does not provide a relative state of charge in percent, but it is what it is.

Please also add an e.g. `ltc2959-emulation`, for example [emul_lc709203f.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/fuel_gauge/lc709203f/emul_lc709203f.c), [emul_max17048](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/fuel_gauge/max17048/emul_max17048.c), or [emul_bq27z746.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/fuel_gauge/bq27z746/emul_bq27z746.c) and also a test (look [here](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/fuel_gauge)), because this helps a lot to avoid possible future breaking changes and further increase the test coverage and quality of Zephyr.

LostinTimeandspaceYT commented on drivers/fuel_gauge/ltc2959/ltc2959.h:
```diff
@@ -0,0 +1,187 @@
+/* Used when ACR is controlled via firmware */
+#define LTC2959_ACR_CLR (0xFFFFFFFF)
+
+#ifdef CONFIG_FUEL_GAUGE_LTC2959_RSENSE_MOHMS
```
Thank you, these are helpful examples. I will add an `emul` and ztest.

LostinTimeandspaceYT commented on dts/bindings/fuel-gauge/adi,ltc2959.yaml:
```diff
@@ -0,0 +1,24 @@
+    description: |
+      Optional GPIO used for ALERT output. The LTC2959 pulls this pin low
+      when thresholds are crossed or charge events occur. Typically wired
+      to a host interrupt-capable GPIO pin.
```
I ended up removing this from the binding for the sake of simplicity. I am happy to add support for this hardware feature in the future.

LostinTimeandspaceYT commented:
@ttmut , I moved the `RSENSE` from Kconfig to a DeviceTree binding, removed the `label` and `alert-gpios` from the binding.

@DBS06 The commit history should be all set. I separated it into 3 commits, one for the driver, one for the custom properties, and the last for the test/emulation code. 

Thanks again for your folks' help.

DBS06 commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,544 @@
+{
+	uint8_t reg = high ? LTC2959_REG_TEMP_THRESH_HIGH_MSB : LTC2959_REG_TEMP_THRESH_LOW_MSB;
+	uint16_t raw = ((uint32_t) value_dK << 16) / LTC2959_TEMP_K_SF;
+	uint8_t buf[2] = {raw >> 8, raw & 0xFF};
```
Better use `sys_put_be16()` or from `byteorder.h`
```suggestion
	uint8_t buf[2];
	sys_put_be16(raw, buf);
```

DBS06 commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,544 @@
+	uint8_t reg = high ? LTC2959_REG_VOLT_THRESH_HIGH_MSB : LTC2959_REG_VOLT_THRESH_LOW_MSB;
+	const uint16_t raw = (((uint64_t) value << 15) / LTC2959_VOLT_THRESH_UV_SCALAR);
+
+	uint8_t buf[2] = {raw >> 8, raw & 0xFF};
```
Better use `sys_put_be16()` or from `byteorder.h`
```suggestion
	uint8_t buf[2];
	sys_put_be16(raw, buf);
```

DBS06 commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,544 @@
+		case LTC2959_GPIO_MODE_BIPOLAR: {
+			int16_t raw = (int16_t) (((int64_t) value_uv * 32768) /
+			        LTC2959_GPIO_BIPOLAR_UV_SF);
+			uint8_t buf[2] = {raw >> 8, raw & 0xFF};
```
Better use `sys_put_be16()` or from `byteorder.h`
```suggestion
			uint8_t buf[2];
			sys_put_be16(raw, buf);
```

DBS06 commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,544 @@
+		case LTC2959_GPIO_MODE_UNIPOLAR: {
+			uint16_t raw = (uint16_t) (((int64_t) value_uv * 32768) /
+			        LTC2959_GPIO_UNIPOLAR_UV_SF);
+			uint8_t buf[2] = {raw >> 8, raw & 0xFF};
```
Better use `sys_put_be16()` or from `byteorder.h`
```suggestion
			uint8_t buf[2];
			sys_put_be16(raw, buf);
```

DBS06 commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,544 @@
+
+			uint8_t gpio_mode = adc_mode & LTC2959_CTRL_GPIO_MODE_MASK;
+			if (gpio_mode == LTC2959_GPIO_MODE_CHGCOMP) {
+				// NOTE: In CHGCOMP mode, ACR is cleared by pulling GPIO low
+				// externally
```
Note: C99 comments are not allowed by the "Compliance Check"

DBS06 commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,544 @@
+			break;
+
+		case FUEL_GAUGE_LOW_VOLTAGE_ALARM:
+			ret = ltc2959_set_voltage_threshold_uv(dev, false, val.voltage);
```
Wrong `fuel_gauge_prop_val`
```suggestion
			ret = ltc2959_set_voltage_threshold_uv(dev, false, val.low_voltage_alarm);
```

DBS06 reviewed (changes requested):
All in all your PR looks very good and mature!

The requested changes from my side are mostly minor issues.  
As mentioned [here](https://github.com/zephyrproject-rtos/zephyr/pull/90356#discussion_r2241559895), consider adding these properties:
```
	FUEL_GAUGE_HIGH_VOLTAGE_ALARM,
	FUEL_GAUGE_HIGH_CURRENT_ALARM,
	FUEL_GAUGE_LOW_CURRENT_ALARM,
	FUEL_GAUGE_HIGH_TEMPERATURE_ALARM,
	FUEL_GAUGE_LOW_TEMPERATURE_ALARM,
```
to `enum fuel_gauge_prop_type` in `fuel_gauge.h` and add the corresponding `union fuel_gauge_prop_val`, and, as I mentioned, don't forget to add a documentation for your custom properties.

LostinTimeandspaceYT commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,544 @@
+static int ltc2959_write_acr(const struct device* dev, uint32_t value)
+{
+	const struct ltc2959_config* cfg = dev->config;
+	uint8_t buf[4] = {
+	        (value >> 24) & 0xFF, (value >> 16) & 0xFF, (value >> 8) & 0xFF, value & 0xFF};
```
I will resolve each of these by using the `sys_put_beXX` function.

DBS06 commented:
For your information: with the PR #93836 (which will most likely be merged earlier) there will be slight change in the `CMakeLists.txt`, please adapt to this after the PR is merged.

LostinTimeandspaceYT committed:
drivers: fuelgauge: Added properties to prop_type.

Adds properties to fuel gauge api to support ADI LTC2959.

Signed-off-by: Nathan Winslow <natelostintimeandspace@gmail.com>

LostinTimeandspaceYT committed:
drivers: fuelgauge: Added test for ADI LTC2959

Adds test and native sim support for ADI LTC2959 fuel gauge.

Signed-off-by: Nathan Winslow <natelostintimeandspace@gmail.com>

LostinTimeandspaceYT commented on drivers/fuel_gauge/ltc2959/ltc2959.h:
```diff
@@ -0,0 +1,135 @@
+	LTC2959_CC_DEADBAND_40UV = (0b10 << 6),
+	LTC2959_CC_DEADBAND_80UV = (0b11 << 6),
+	LTC2959_CC_DO_NOT_COUNT = BIT(3),
+};
```
done

LostinTimeandspaceYT commented:
@DBS06 I modified the `CMakelists.txt` in preparation for #93836 . I added the properties to `fuel_gauge.h`, and made the changes to driver as requested. Additionally, I removed the `FUEL_GAUGE_CC_HOLD` and `FUEL_GAUGE_CC_CLEAR` properties in favor of exposing api functions to handle that functionality. I felt this made more sense as its behavior specific to this IC and not a property perse.

dsseng commented on tests/drivers/fuel_gauge/ltc2959/test_ltc2959.c:
```diff
@@ -0,0 +1,111 @@
+	ret = fuel_gauge_get_prop(fixture->dev, FUEL_GAUGE_LOW_VOLTAGE_ALARM, &val1);
+
+	zassert_equal(ret, 0, "Failed to get low voltage threshold");
+	LOG_INF("low voltage threshold: %d µV", val1.voltage);
```
Maybe also assert val == val1 here to make sure reading works fine?

dsseng commented on drivers/fuel_gauge/ltc2959/emul_ltc2959.c:
```diff
@@ -0,0 +1,266 @@
+	data->regs[LTC2959_REG_GPIO_THRESH_HIGH_MSB] = 0x7F;
+	data->regs[LTC2959_REG_GPIO_THRESH_HIGH_LSB] = 0xFF;
+	data->regs[LTC2959_REG_GPIO_THRESH_LOW_MSB] = 0x80;
+
```
Maybe provide sample data for things like accumulated charge here to test the decoding and calculations the driver does?

LostinTimeandspaceYT replied:
I added some additional tests to this most recent change. Notable tests to check the new property types, checks for register sanitizing, and a test for the overflow/underflow behavior of the ACR. If additional tests are required, I am happy to add them and amend the commit.

dsseng commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,615 @@
+
+struct ltc2959_config {
+	struct i2c_dt_spec i2c;
+	uint32_t rsense_mohms;
```
Seems to be unused, `current_lsb_uv` is calculated from the DT property directly and `rsense_mohms` is never read

dsseng replied:
That makes sense, thanks

LostinTimeandspaceYT replied:
changes applied.

DBS06 reviewed (dismissed):
Well done! From my side this PR is finished!

LostinTimeandspaceYT commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,629 @@
+static inline uint32_t ltc2959_counts_to_uah(uint32_t counts)
+{
+	/* µAh = round(counts * 533 / 1000) */
+	uint64_t prod = (uint64_t)counts * (uint64_t)LTC2959_ACR_UAH_NUM;
```
I don't believe the subtraction is necessary since the result gets rounded, but I believe you may be correct about the scaling.  I can verify on a target tomorrow.

LostinTimeandspaceYT replied:
My apologizes for the delayed response. I see what you are saying, yes that makes sense. I will work on a fix and patch it. Thank you for clarifying.

perencia-wc commented on drivers/fuel_gauge/ltc2959/ltc2959.c:
```diff
@@ -0,0 +1,629 @@
+			return ret;
+		}
+
+		val->remaining_capacity = ltc2959_counts_to_uah(acr);
```
Isn't this the spent/increased capacity, instead of the remaining?

DBS06 reviewed (dismissed):
You need to resolve the conflict, otherwise so far it looks to me that you addressed all requested changes.   
Well Done :+1:

aaronemassey reviewed (dismissed):
Thank you for adding tests/ emulator!
